src/equations/shallow_water_wet_dry_2d.jl

```diff
@@ -242,6 +241,302 @@ Should be used together with [`Trixi.TreeMesh`](@extref).
+    g = equations.gravity
+
+    # Get the water height and velocity from the inner state
+    h_inner = waterheight(u_inner, equations)
+    v1_inner, v2_inner = velocity(u_inner, equations)
+
+    # Extract the external momentum from the boundary condition
+    hv1_boundary, hv2_boundary = boundary_condition.hv_boundary(t)
+
+    # Calculate the boundary state based on the direction.
+    # To extrapolate the external water height `h_boundary` assume that the Riemann invariant remains 
+    # constant across the incoming characteristic. This requires to solve for the roots of a nonlinear function.
+    # For convenience we substitute x = h_boundary and solve for x.
+    if direction == 1 # x-
+        fx = ZeroProblem(x -> 2 * sqrt(g) * x^(3 / 2) +
+                              (v1_inner - 2 * sqrt(g * h_inner)) * x - hv1_boundary,
+                         h_inner)
+        h_boundary = solve(fx, Order2())
+        if hv1_boundary > 0
+            u_boundary = SVector(h_boundary, hv1_boundary, hv2_boundary, u_inner[4])
+        else
+            u_boundary = SVector(h_boundary, hv1_boundary, u_inner[3], u_inner[4])
+        end
+    elseif direction == 2 # x+
+        fx = ZeroProblem(x -> 2 * sqrt(g) * x^(3 / 2) -
+                              (v1_inner + 2 * sqrt(g * h_inner)) * x + hv1_boundary,
+                         h_inner)
+        h_boundary = solve(fx, Order2())
+        if hv1_boundary < 0
+            u_boundary = SVector(h_boundary, hv1_boundary, hv2_boundary, u_inner[4])
+        else
+            u_boundary = SVector(h_boundary, hv1_boundary, u_inner[3], u_inner[4])
+        end
+    elseif direction == 3 # y-
+        fx = ZeroProblem(x -> 2 * sqrt(g) * x^(3 / 2) +
+                              (v2_inner - 2 * sqrt(g * h_inner)) * x - hv2_boundary,
+                         h_inner)
+        h_boundary = solve(fx, Order2())
+        if hv2_boundary > 0
+            u_boundary = SVector(h_boundary, hv1_boundary, hv2_boundary, u_inner[4])
+        else
+            u_boundary = SVector(h_boundary, u_inner[2], hv2_boundary, u_inner[4])
+        end
+    elseif direction == 4 # y+
+        fx = ZeroProblem(x -> 2 * sqrt(g) * x^(3 / 2) -
+                              (v2_inner + 2 * sqrt(g * h_inner)) * x + hv2_boundary,
+                         h_inner)
+        h_boundary = solve(fx, Order2())
+        if hv2_boundary < 0
+            u_boundary = SVector(h_boundary, hv1_boundary, hv2_boundary, u_inner[4])
+        else
+            u_boundary = SVector(h_boundary, u_inner[2], hv2_boundary, u_inner[4])
+        end
+    end
+
+    # Evaluate the conservative flux at the boundary
+    flux = Trixi.flux(u_boundary, orientation, equations)
+
+    # Return the conservative and nonconservative fluxes. 
+    # The nonconservative part is zero as we assume a constant bottom topography at the boundary.
+    return (flux, zero(u_inner))
+end
+
+# Version for `StructuredMesh`, `UnstructuredMesh2D` and `P4estMesh`
```

Comment:
Similar to comment above. This might not work on `StructuredMesh` due to orientation of the normals.